### ISSUE TITLE
Unable to run under windows

### ISSUE THREAD

**Issue by AndreMPCosta** [philippj/SteamworksPy#1]
First let me thank you for doing this for Python.

I'm having some troubles using this under Windows:

C:\Python27\python.exe D:/secretweapon/SteamworksForPython-master/Tests/steam-api-test.py
SteamworksPy loaded for Windows
Traceback (most recent call last):
  File "D:/secretweapon/SteamworksForPython-master/Tests/steam-api-test.py", line 32, in <module>
    S_API.IsSteamRunning.restype = c_bool
  File "C:\Python27\lib\ctypes__init__.py", line 378, in **getattr**
    func = self.**getitem**(name)
  File "C:\Python27\lib\ctypes__init__.py", line 383, in **getitem**
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: function 'IsSteamRunning' not found

Process finished with exit code 1

I'm using the .dll that is on the sdk inside the folder release redistributable_bin, I think that's the one I should be using.

The only access I have is to GetPlayerSteamLevel and GetSteamID.

**Comment by Gramps** [comment]
Hey there!  Welcome!

So, I'll ask a few questions first.  Are you running and logged into Steam when running the tests?  Steam has to be on and you have to be logged into an account or the test will fail.  It looks like SteamworksPy itself loaded fine so we can skip that.

Did you build the SteamworksPy.dll yourself or using the pre-build from the repo?  And yes, the steam_api.dll from redistributable_bin is needed.

**Comment by AndreMPCosta** [comment]
I've tried the pre-build from the repo. I've assumed it was to rename the steam_api.dll from redistributable_bin to SteamworksPy.dll, guess this was my mistake :P 
So probably I'll have to compile by myself using the lib, right?

**Comment by Gramps** [comment]
You can use the pre-build from the repo; it doesn't need renamed.  But you also need the steam_api.dll from redistributable_bin.  Both are required for it to work.

You can compile it from scratch if you want but it shouldn't be necessary.

Just make sure the SteamworksPy.dll and the steam_api.dll are in the base folder with the test your are running or your game's base.

**Comment by AndreMPCosta** [comment]
SteamworksPy loaded for Windows
Traceback (most recent call last):
  File "D:/secretweapon/SteamworksForPython-master/Tests/steam-api-test.py", line 36, in <module>
    S_API.RequestAppProofOfPurchaseKey.restype = c_char_p
  File "C:\Python27\lib\ctypes__init__.py", line 378, in **getattr**
    func = self.**getitem**(name)
  File "C:\Python27\lib\ctypes__init__.py", line 383, in **getitem**
    func = self._FuncPtr((name_or_ordinal, self))
AttributeError: function 'RequestAppProofOfPurchaseKey' not found

I guess this one isn't working.

**Comment by Gramps** [comment]
Oh shit... haha, okay, this is my fault actually.  The DLL files in the pre-builds are from our last version and haven't been recompiled to use the newest version of SteamworksPy!  I only updated the Linux SO file!

You will have to compile the SteamworksPy.dll files from scratch actually.  As will I so I can update the pre-build folders.  Thanks for reminding me of this!

**Comment by AndreMPCosta** [comment]
Yes, some functions aren't working I've commented them and the file at least runs. Guess I'll have to compile the .dll then, wish me luck! :P

**Comment by Gramps** [comment]
Good luck!  I added a note about the pre-builds in the read me.

Yeah, compiling them in Windows is a lot more involved than in Linux.  Linux is just a simple command line but I could never get them to compile in Windows that way.  Always had to do the Visual Studio stuff.

Also note that you should ship Visual C++ 2015 Redistributable with your game as some players will need to install that.

And since we are talking about it, I might as well try compiling it now.

**Comment by AndreMPCosta** [comment]
Can't find the .cpp file (I'm following your guide):

Add steampy.cpp to Source Files and steam_api.h from /steam/ folder to Header Files

Can you point me where it is? :)

**Comment by Gramps** [comment]
Steampy.cpp is renamed SteamworksPy.cpp in the repo.  I should update that in the read me too!  I'm slipping!

**Comment by AndreMPCosta** [comment]
I've compiled successfully, now every function is working. Thank you for your help Gramps!

**Comment by AndreMPCosta** [comment]
User test is not working though:

--- Steamworks API Test -------------------------------------
Which test do you want to use?
(A)pps, (F)riends, (M)usic, (U)ser, User (S)tats, U(t)ilities, (Q)uit
Run which test?: U

.....running user test.....
Steam ID: 0
Steam Level: 0
.....user test done.....

**Comment by Gramps** [comment]
Glad to hear you got it going!

Hmm, I know the Steam ID thing never works.  Has to do with how they encode the number... I was never able to get it to come out correctly.  Steam Level should have worked but might also be faulty in this case.  Thanks for letting me know and I'll check into them soon.

They just updated Steamworks to 1.35a so I'll be updating the library again.  Maybe work out the kinks with these two!

**Comment by AndreMPCosta** [comment]
Yeah I'll be watching the progress! Thank you again for the development and for your help! 
I'm developing in Panda3d and it would be awesome if I could try to "green light" the project :)

**Comment by Gramps** [comment]
Cheers!  Yeah, I hope to get some more features built in.  Since it is a wrapper of the API there area, sadly, some things that just can't be as far as I can tell.

Nice, feel free to send me a link to your game if/when you get it on Steam Greenlight!

**Comment by AndreMPCosta** [comment]
By the way, I've used a Wrapper for Delphi some time ago:

https://github.com/thecocce/steamwrapper

I don't know if it can help you with something, this wrapper worked flawlessly as far as I remember.

**Comment by Gramps** [comment]
I'll check it out!  Thanks for the tip.

**Comment by AndreMPCosta** [comment]
Re-Used the delphi wrapper posted before and look what I got ;)

S_API.System_InitWrapper()
print 'User Level: ' + str(S_API.SteamUser_GetPlayerSteamLevel())
print 'User ID: ' + str(S_API.SteamUser_GetSteamID())
print 'Number of Friends : ' + str(S_API.SteamFriends_GetFriendCount())

C:\Python27\python.exe D:/secretweapon/SteamworksForPython-master/Tests/api-test-2.py
libSteamWrapper loaded for Windows
User Level: 6
User ID: 6224876
Number of Friends : 56
Setting breakpad minidump AppID = 480
Steam_SetMinidumpSteamID:  Caching Steam ID:  76561197990283346 [API loaded no]

Process finished with exit code 0

**Comment by Gramps** [comment]
Haha, interesting.  I wonder if that Steam ID is right, or how to even find your actual Steam ID.  I have gotten ID responses from some tests but the numbers always look wrong compared to what I think my ID number actually is.

I can't remember if the Steam ID issue is from the CPP file or Steamworks.py file...

But that is interesting.  I will dig into it now actually.

**Comment by AndreMPCosta** [comment]
I've converted it to 64 bits but it's wrong:

C:\Python27\python.exe D:/secretweapon/SteamworksForPython-master/Tests/api-test-2.py
libSteamWrapper loaded for Windows
User Level: 6
User ID: 76561197966490604
Number of Friends : 56
Setting breakpad minidump AppID = 480
Steam_SetMinidumpSteamID:  Caching Steam ID:  76561197990283346 [API loaded no]

The DumpID is right though, idk how I can catch that ID :P 

The problem probably lies here:
Severity    Code    Description Project File    Line
Warning C4190   'SteamServer_GetSteamID' has C-linkage specified, but returns UDT 'CSteamID' which is incompatible with C   Win32Project3   d:\secretweapon\steamworksforpython-master\src\csteamgameserver.cpp 54

It's expecting a C type thing (I'm not very experienced here)

**Comment by Gramps** [comment]
Yeah, I think SteamworksPy gets an error during compiling on that as, I believe, the Steam ID is int64(?).  I don't really understand the function in the Steam API as instead of bool, void, int, etc. it is shown as CSteamID which is called in other functions through the API.

Plus I'm not very well versed in C++....  :\

**Comment by AndreMPCosta** [comment]
Isn't it possible to wrap the steam_api_flat?

**Comment by Gramps** [comment]
I remember trying it at one point and abandoned it... but I don't quite remember why.  This was, of course, at the beginning of the whole thing and maybe now I would be able to get it working right.

**Comment by AndreMPCosta** [comment]
If you could that, would be awesome. 

Their notes:
// Purpose: Header for flatted SteamAPI. Use this for binding to other languages.
// This file is auto-generated, do not edit it.

I'm trying to do it, not having much success though.

The source that I've got from delphi wrapper works ok for stuff that it's in C, I have like 13 functions that doesn't behave like they should, not terrible. The problem it's that the SteamID is critical for all these functions.

**Comment by Gramps** [comment]
Yeah, the SteamID seems to be the most pain in the ass one for some reason.  I am looking over the flat API now.  I am going to make some small test and try to compile an SO from it.

**Comment by AndreMPCosta** [comment]
Good Luck then, fingers crossed!

**Comment by Gramps** [comment]
Indeed!  Should be pretty quick.  Just going to try a few simple things I know should work with no problems.

**Comment by AndreMPCosta** [comment]
The thing that is bugging me out it the dump getting the right ID, I can't figure out how the ID is extracted (at cSteamSystem). Understanding that would solve almost all my issues with the delphi port.

**Comment by Gramps** [comment]
Yeah, the ID is kind of confusing.  I remember reading something about it a while back, mostly about how the number is stored and decoded(?).  Can't remember though.

Looking at the Flat API, there seems to be no way to initialize Steam like in the normal API.  That or they renamed it something wholly different.

**Comment by AndreMPCosta** [comment]
I think it's this one: S_API void SteamAPI_ISteamClient_SetWarningMessageHook(intptr_t instancePtr, SteamAPIWarningMessageHook_t pFunction);

Not sure though

**Comment by Gramps** [comment]
Hmm. Maybe?  I'll try it and see what happens.

**Comment by Gramps** [comment]
I get TONS of these errors:
`error: ‘S_API’ does not name a type`
Which is the beginning of every line in the Flat API.  Not sure if this is meant to work in conjunction with the normal API or I'm missing something.  But the Flat APi never declares the S_API name

**Comment by AndreMPCosta** [comment]
By the way more info about the problem: 
One thing that was not supported was classes. Object models of different languages were too different, and for system programming Microsoft assumed you'd be using C in any case. Full support for object-oriented programming didn't come around until .NET, and then only for .NET style objects and classes.

The practical upshot of this is that while you can have a class inside of a DLL, you have to 'unwrap' the class by providing external functions which act as an intermediary to the class's methods.

So basically we would have to extract the class attributes and methods from CSteamID.

For flat api,
I defined S_API as #define S_API extern "C" __declspec( dllexport ), but didn't work, I just can't call any function in python, it's goes beyond my knowledge :P

**Comment by Gramps** [comment]
Yeah, the CSteamID seemed like it required a few more steps that I couldn't do with a wrapper in Python.  Perhaps coding something into the .CPP file prior to compiling?

Haha, I think that is why I gave up on the Flat API previously.  I may have an old test lying around in my Tests folder but I doubt it'd help at all.  I'll tinker with it for another 30 minutes and see if I get anywhere.

**Comment by AndreMPCosta** [comment]
Hoorrayyy:

C:\Python27\python.exe D:/secretweapon/SteamworksForPython-master/Tests/api-test-2.py
libSteamWrapper loaded for Windows
User Level: 6
User ID: 76561197990283346
Number of Friends : 56
Setting breakpad minidump AppID = 480
Steam_SetMinidumpSteamID:  Caching Steam ID:  76561197990283346 [API loaded no]

**Comment by Gramps** [comment]
Is this from the Delphi wrapper?

**Comment by AndreMPCosta** [comment]
Yeap :)

**Comment by Gramps** [comment]
Interesting... and that is for D?

**Comment by AndreMPCosta** [comment]
I didn't understand the question :P 
Trying to fix the instances that use CSteamID, at least that one is working.

**Comment by Gramps** [comment]
That's good.  What part of that are you using?

**Comment by AndreMPCosta** [comment]
I'm using their sources, removed Virtual Reality and other module that was causing me trouble. Probably the callbacks won't work, but that's not critical to be honest. There are lots of methods available.

**Comment by Gramps** [comment]
Hmm, I'll have to dig into that more then.  I am still riding the Flat API wave and getting a little progress...

**Comment by AndreMPCosta** [comment]
Well when I finish I can share what I've got so far :) 
Good Luck with Flat!

**Comment by Gramps** [comment]
Excellent. Look forward to checking it out!  Yeah, I hope my tests pay off because it should make all this a lot easier!

**Comment by AndreMPCosta** [comment]
Any tips how to convert directly to 64 bits in C?

I always have to do the "add" on python.

SW_PY uint64 SteamUser_GetSteamID_GetAccountID() {return static_cast '<'long long'>'(SteamUser()->GetSteamID().GetAccountID() + 76561197960265728); };

I'm probably doing something wrong, not very important, but it would be nice to return 64 bits and no need to convert in python.

**Comment by Gramps** [comment]
That is a good question... I'm not too well versed in C actually.  Hence building games in Python (just easier to build and learn).

So you want to return the 64 bit back to the API system or to something else?  Where is it going?

I'm still trying to print a pointer into something that I can human read.

**Comment by AndreMPCosta** [comment]
I want to return the actual SteamID, which is 64 bits so we need to add 76561197960265728. It works fine in Python, but I can't manage to do this in C, so I don't need to do the conversion in Python.

**Comment by Gramps** [comment]
Ah.  That I am at a loss actually.  Not sure on how to do that in C, sadly.

**Comment by Gramps** [comment]
I was trying to figure that out a while ago actually!  If you get it, that'll solve a few problems!

**Comment by AndreMPCosta** [comment]
I'll leave at as it is, it's good for now. Python does the conversion :P

**Comment by Gramps** [comment]
Fair enough.  What did you use in Python?

**Comment by AndreMPCosta** [comment]
Just need to add that big number 76561197960265728 :)

**Comment by Gramps** [comment]
Hmm, that the Steam_SetMinidumpSteamID:  Caching Steam ID?

**Comment by AndreMPCosta** [comment]
Nope, I mean it's the same number, but it's obtained differently. I call the function I posted earlier which returns me the ID in 32 bits. Then I convert it to 64 bits doing the + 76561197960265728.

**Comment by Gramps** [comment]
Oh, gotcha.

**Comment by AndreMPCosta** [comment]
Basically I figured out this using an old Delphi file:

  TSteamID = record
  private
    Data1: int32; // 32 bit
    Data2: SmallInt; // 16 bit
    Data3: Byte; // 8 bit
    Data4: Byte; // 8 bit
  public
    function GetAccountID: uint32; // unique account identifier
    function GetAccountInstance: uint32; // dynamic instance ID
    function GetAccountType: TAccountType; // type of account
    function GetUniverse: TUniverse; // universe this account belongs to
  end;

**Comment by AndreMPCosta** [comment]
I can expose these 4 methods, I'm just exposing the 1st one because that's the more relevant one.

**Comment by Gramps** [comment]
I see.  Looks like a bit of reverse-engineering is in order to get this all sorted out.  Would be great to have the SteamID functioning correctly so we can use other functions.

**Comment by AndreMPCosta** [comment]
It's figured out, it's a pain in the ass to convert all the instances of the class CSteamID to the exposed method. I'm just converting the problematic outputs, inputs its another story, not for today at least :P

**Comment by Gramps** [comment]
Haha, yes, there are always other days.  Cheers for getting that all figured out!

**Comment by AndreMPCosta** [comment]
Would be nice to have the major features working for Python. We'll try :)

**Comment by Gramps** [comment]
Yeah, we mostly focused on what we needed at the time with notes on expanding out to additional features later on.  Most of the features are fairly easy to get working but the SteamID was a hang up for a while.

Ideally it would be great to port everything over.

**Comment by AndreMPCosta** [comment]
Okay I've finished all the functions that don't need CSteamID as input. Those will be harder to get working, not impossible I guess.

**Comment by Gramps** [comment]
Damn!  That was quick!  Yeah, I don't think they'll be impossible.  Usually just requires clearing a hurdle or two.

**Comment by AndreMPCosta** [comment]
I'm having a problem with this, I can make calls but I can't assign nothing to a variable, maybe you have an idea why:

libSteamWrapper loaded for Windows
  File "D:/secretweapon/SteamworksForPython-master/Tests/api-test-2.py", line 33, in <module>
    friends_count = S_API.SteamFriends_GetFriendCount()
WindowsError: exception: access violation reading 0x00000000

**Comment by Gramps** [comment]
Hmm, I have seen this before.  Are you using the Delphi wrapper or the SteamworksPy wrapper?

Is this from one of the tests we have in the test folder?

I believe the Steam API name you are calling is incorrect...

**Comment by AndreMPCosta** [comment]
Delphi.

S_API.SteamUser_GetSteamID_GetAccountID.restype = c_uint64
user_id = S_API.SteamUser_GetSteamID_GetAccountID() + 76561197960265728
print user_id

:(

**Comment by Gramps** [comment]
This is pulling from the normal Steam API?  Using only the steam_api.dll?

**Comment by AndreMPCosta** [comment]
Yes, their source uses normal Steam API.

**Comment by Gramps** [comment]
They have a compiled DLL their library creates?

**Comment by AndreMPCosta** [comment]
They have, but I created my own with their sources. Yesterday all my comments were using this compiled dll from their sources.

**Comment by Gramps** [comment]
Gotcha.  Yeah, I just downloaded and opened their repo.  So the other functions seem to work, just not this one?

**Comment by AndreMPCosta** [comment]
The functions work, I just can't assign anything to a variable. For example if I print the call without assigning it, works fine.

**Comment by Gramps** [comment]
You know, I had this same issue with Flat API.  It works fine if not passed to a variable but causes a segmentation fault if it is.  I wasn't totally sure what that was about.  However, with the Flat API it just returns a function pointer and nothing human readable.  Though I think I was failing to make it readable from the pointer.  But realized after an hour that I was versed enough in C to make that happen currently.

I'm digging into the Delphi thing real quick to see if I can figure out why.  SteamworksPy has no problem passing to variables but what is the difference between the two?

**Comment by Gramps** [comment]
Does friend count work without passing it to a variable?

**Comment by AndreMPCosta** [comment]
Yes!

**Comment by Gramps** [comment]
Haha, son of.  Well, I think if we solve this I may be able to get the Flat API working.

**Comment by Gramps** [comment]
Can you try something.  Try this:
`friends_count = S_API.SteamFriends_GetFriendCount`
...instead of...
`friends_count = S_API.SteamFriends_GetFriendCount()`

**Comment by Gramps** [comment]
For whatever reason this works okay with Flat API and allows me to pass it to a variable no problem.  However, with Flat API I still get the pointer reference and not anything I can work with.  For example friend count doesn't pass an readable integer, just a pointer.

**Comment by AndreMPCosta** [comment]
That gives me a pointer if I don't use ()

**Comment by Gramps** [comment]
Hmm... so you get a pointer with () and a crash without?   This is the same issue I have with the Flat API.

**Comment by Gramps** [comment]
You wouldn't know how to convert that pointer to something readable, would you?

**Comment by AndreMPCosta** [comment]
I'm dumb as fuck, I wasn't calling this S_API.System_InitWrapper() before assigning stuff :facepalm:

Now works as it should.

**Comment by Gramps** [comment]
Haha. Ohhhh... wait a second.  Let me try something.

**Comment by AndreMPCosta** [comment]
S_API.System_InitWrapper()
user_id = S_API.SteamUser_GetSteamID_GetAccountID() + 76561197960265728
print 'User Level: ' + str(S_API.SteamUser_GetPlayerSteamLevel())
print 'User ID: ' + str(user_id)
print 'Number of Friends : ' + str(S_API.SteamFriends_GetFriendCount())
while 1:
    pass

C:\Python27\python.exe D:/secretweapon/SteamworksForPython-master/Tests/api-test-2.py
libSteamWrapper loaded for Windows
User Level: 6
User ID: 76561197990283346
Number of Friends : 56

**Comment by Gramps** [comment]
That is nice.  Still hasn't solved my Flat API issues, but is a step in the right direction!  Cheers.

**Comment by Gramps** [comment]
You ever figure out overlay?

**Comment by AndreMPCosta** [comment]
I didn't try anything else in the meantime, not sure if overlay works, I didn't try.

**Comment by Gramps** [comment]
I actually got overlay working fine but it has to be run from Steam directly and not from terminal or any other method.  It also requires, as I mentioned in the readme, OpenGL or D3D.

**Comment by Gramps** [comment]
Never closed this for some reason.  After all this time I'm going to pull over a test for the Steam ID from the API system I have for Godot!  Should be in the next update.
